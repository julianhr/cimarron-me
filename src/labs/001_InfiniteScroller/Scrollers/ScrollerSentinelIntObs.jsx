import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: auto;
`

export class ScrollerSentinelIntObs extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func.isRequired,
    isFetching: PropTypes.bool.isRequired,
    sentinelPosition: PropTypes.number.isRequired,
    setEntryCount: PropTypes.func.isRequired,
    setIsFetching: PropTypes.func.isRequired,
  }

  state = {
    isSupported: true,
    cards: [],
    fetch: {
      status: 'loading',
      error: null,
    },
  }

  refSentinel = React.createRef()
  observer = null

  componentDidMount() {
    this.props.setEntryCount(0)

    if (this.isSupported()) {
      this.observer = new IntersectionObserver(this.handleIntObs())
      this.getNewCards()
    } else {
      this.setState({ isSupported: false, })
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { isSupported, cards } = this.state

    if (isSupported && cards.length > prevState.cards.length) {
      this.observer.observe(this.refSentinel.current)
    }
  }

  isSupported() {
    // Boolean check needed for Node and Jest by extension
    return 'IntersectionObserver' in window && Boolean(IntersectionObserver)
  }

  async getNewCards() {
    const { cardFetcher, setIsFetching, setEntryCount } = this.props

    try {
      setIsFetching(true)
      const data = await cardFetcher(this)
      this.setNewCards(data)
      setEntryCount(this.state.cards.length)
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
    }

    setIsFetching(false)
  }

  getCard(datum, i, sentinelPosition, lastKey) {
    const { title, image_url: imgUrl, description } = datum
    const position = lastKey + i + 1
    const props = { imgUrl, title, description, position }
    let CardElement = Card

    if (i + 1 === sentinelPosition) {
      CardElement = React.forwardRef((props, ref) => <Card forwardedRef={ref} {...props} />)
      props.forwardedRef = this.refSentinel
    }

    return <CardElement key={lastKey + i} {...props} />
  }

  setNewCards(data) {
    const { length: lastKey } = this.state.cards
    const { sentinelPosition } = this.props
    const cards = [...this.state.cards]

    // a new sentinel will be set, remove the previous one
    if (this.refSentinel.current) {
      this.observer.unobserve(this.refSentinel.current)
    }

    data.forEach((datum, i) => {
      const card = this.getCard(datum, i, sentinelPosition, lastKey)
      cards.push(card)
    })

    this.setState({ cards, fetch: null })
  }

  handleIntObs() {
    const parent = this

    return (entries) => {
      entries.forEach(entry => {
        const { isFetching } = parent.props
        const { cards } = parent.state

        if (entry.isIntersecting && !isFetching && cards.length < 200) {
          parent.getNewCards()
        }
      })
    }
  }

  renderResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (!this.state.isSupported) {
      return (
        <FetchStatus>
          <>
            <p>This browser doesn't support the Intersection Observer.</p>
            <a
              href='https://caniuse.com/#feat=intersectionobserver'
              target='_blank'
              rel='noopener noreferrer'
            >
              Check compatibility
            </a> 
          </>
        </FetchStatus>
      )
    } else if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error' && this.state.cards.length === 0) {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root>
        {this.renderResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ sentinelPosition, isFetching }) => (
  { sentinelPosition, isFetching }
)

const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerSentinelIntObs)
