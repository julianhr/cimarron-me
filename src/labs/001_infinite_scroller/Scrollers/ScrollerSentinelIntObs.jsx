import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import Card from '../Card'
import { setEntryCount, setIsFetching } from '../actions/rootActions'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: auto;
`

const Loading = styled.div`
  background: #eee;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 20px;
  padding: 40px;
`

class ScrollerSentinelIntObs extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func,
    isFetching: PropTypes.bool,
    sentinelPosition: PropTypes.number,
    entryCount: PropTypes.number,
    setEntryCount: PropTypes.func,
    setIsFetching: PropTypes.func,
  }

  state = {
    isSupported: false,
    cards: [],
    loading: {
      status: 'loading',
      error: null,
    },
  }

  refSentinel = React.createRef()
  observer = null

  componentDidMount() {
    if (this.isSupported()) {
      this.observer = new IntersectionObserver(this.handleIntObs())
      this.setState({ isSupported: true })
      this.fetchCards()
    } else {
      this.setState({ isSupported: false, })
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { isSupported, cards } = this.state

    if (isSupported && cards.length > prevState.cards.length) {
      this.observer.observe(this.refSentinel.current)
    }
  }

  isSupported() {
    return 'IntersectionObserver' in window
  }

  fetchCards() {
    const { cardFetcher } = this.props

    cardFetcher()
      .then(data => {
        const { cards: currCards } = this.state
        const cards = [...currCards]

        if (this.refSentinel.current) {
          this.observer.unobserve(this.refSentinel.current)
        }

        this.appendNewCards(cards, data, currCards.length)
        this.setState({ cards, loading: { status: 'success', error: null } })
        this.props.setEntryCount(cards.length)
        this.props.setIsFetching(false)
      })
      .catch(error => {
        this.setState({ loading: { status: 'error', error: error.toString() } })
        this.props.setIsFetching(false)
        console.error('Fetch error:', error)
      })
  }

  appendNewCards(cards, data, lastKey) {
    const { sentinelPosition } = this.props

    data.forEach(({ title, image_url: imgUrl, description }, i) => {
      const position = lastKey + i + 1
      const props = { imgUrl, title, description, position }
      let CardComponent = Card

      if (i+1 === sentinelPosition) {
        CardComponent = React.forwardRef((props, ref) => <Card forwardedRef={ref} {...props} />)
        props.forwardedRef = this.refSentinel
      }

      cards.push( <CardComponent key={lastKey+i} {...props} /> )
    })
  }

  handleIntObs() {
    const parent = this

    return (entries) => {
      entries.forEach(entry => {
        const { isFetching } = parent.props
        const { cards } = parent.state

        if (entry.isIntersecting && !isFetching && cards.length < 200) {
          parent.props.setIsFetching(true)
          parent.fetchCards()
        }
      })
    }
  }

  renderLoadingMsg() {
    if (!this.state.isSupported) {
      return (
        <div>
          <p>This browser doesn't support the Intersection Observer. </p>
          <a
            href='https://caniuse.com/#feat=intersectionobserver'
            target='_blank'
            rel='noopener noreferrer'
          >
            Check compatibility
          </a> 
        </div>
      )
    }

    switch(this.state.loading.status) {
      case 'loading':
        return 'Loading...'
      case 'error':
        return this.state.loading.error
    }
  }

  renderCardResult() {
    if (this.state.cards.length === 0) {
      return (
        <Loading>
          {this.renderLoadingMsg()}
        </Loading>
      )
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root>
        {this.renderCardResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ sentinelPosition, isFetching, entryCount }) => (
  { sentinelPosition, isFetching, entryCount }
)

const mapDispatchToProps = { setEntryCount, setIsFetching }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerSentinelIntObs)
