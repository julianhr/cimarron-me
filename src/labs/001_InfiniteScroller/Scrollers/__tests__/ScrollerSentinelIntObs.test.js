import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'

import { ScrollerSentinelIntObs } from '../ScrollerSentinelIntObs'


class FakeIntersectionObserverEntry {
  constructor(element) {
    this.entry = element
  }

  get isIntersecting() { return true }
}

class FakeIntersectionObserver {
  constructor(handler) {
    this.handler = handler
    this.entries = []
  }

  observe(element) {
    const entry = new FakeIntersectionObserverEntry(element)
    this.entries.push(entry)
  }

  unobserve(element) { !!element }

  execute() {
    this.handler(this.entries)
  }
}

describe('ScrollerSentinelIntObs', () => {
  const { IntersectionObserver: origIntersectionObserver } = global
  const testProps = {
    cardFetcher: async () => ([]),
    isFetching: false,
    sentinelPosition: 5,
    setEntryCount: () => {},
    setIsFetching: () => {},
  }

  afterEach(() => {
    global.IntersectionObserver = origIntersectionObserver
  })

  describe('props', () => {
    test('cardFetcher is required', () => {
      const props = {...testProps}
      delete props.cardFetcher
      expect(() => render(<ScrollerSentinelIntObs {...props} />)).toThrow()
    })

    test('isFetching is required', () => {
      const props = {...testProps}
      delete props.isFetching
      expect(() => render(<ScrollerSentinelIntObs {...props} />)).toThrow()
    })

    test('sentinelPosition is required', () => {
      const props = {...testProps}
      delete props.sentinelPosition
      expect(() => render(<ScrollerSentinelIntObs {...props} />)).toThrow()
    })

    test('setEntryCount is required', () => {
      const props = {...testProps}
      delete props.setEntryCount
      expect(() => render(<ScrollerSentinelIntObs {...props} />)).toThrow()
    })

    test('setIsFetching is required', () => {
      const props = {...testProps}
      delete props.setIsFetching
      expect(() => render(<ScrollerSentinelIntObs {...props} />)).toThrow()
    })
  })

  describe('#isSupported', () => {
    it('returns true if Interserction Observer supported by the browser', () => {
      global.IntersectionObserver = undefined
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      expect(wrapper.getInstance().isSupported()).toBe(false)
    })

    it('returns false if Intersection Observer not supported by the browser', () => {
      global.IntersectionObserver = FakeIntersectionObserver
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      expect(wrapper.getInstance().isSupported()).toBe(true)
    })
  })

  describe('#getCard', () => {
    const testDatum = {
      image_url: 'http://image.com',
      title: 'test title',
      description: ['sentence one', 'sentence two'],
      position: 4,
    }

    it('returns a non-sentinel Card element if datum index != sentinel index', () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      const instance = wrapper.getInstance()
      const article = document.createElement('article')
      const i = 3, sentinelPosition = 5, lastKey = 10

      instance.refSentinel = article
      expect(instance.getCard(testDatum, i, sentinelPosition, lastKey)).toMatchSnapshot()
    })

    it('returns a sentinel Card element if datum index == sentinel index', () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      const instance = wrapper.getInstance()
      const article = document.createElement('article')
      const i = 3, sentinelPosition = 4, lastKey = 10

      instance.refSentinel = article
      expect(instance.getCard(testDatum, i, sentinelPosition, lastKey)).toMatchSnapshot()
    })
  })

  describe('#setNewCards', () => {
    it('should unobserve element if one is being observed', () => {
      // setup env
      global.IntersectionObserver = FakeIntersectionObserver
      jest.spyOn(ScrollerSentinelIntObs.prototype, 'getCard').mockReturnValue('mock card')
      const spySetState = jest.spyOn(ScrollerSentinelIntObs.prototype, 'setState')
      const spy = jest.spyOn(FakeIntersectionObserver.prototype, 'unobserve')
      const article = document.createElement('element')
      const testData = [1,2,3]
      // render
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      const instance = wrapper.getInstance()
      // test without ref set
      instance.setNewCards(testData)
      expect(spy).not.toHaveBeenCalled()
      // test with ref set
      instance.refSentinel.current = article
      instance.setNewCards(testData)
      expect(spy).toHaveBeenCalledWith(article)
    })

    it('appends new cards to the current ones', () => {
      // setup env
      global.IntersectionObserver = FakeIntersectionObserver
      jest.spyOn(ScrollerSentinelIntObs.prototype, 'getCard').mockImplementation(arg => arg)
      jest.spyOn(ScrollerSentinelIntObs.prototype, 'getNewCards').mockReturnValue(null)
      const spySetState = jest.spyOn(ScrollerSentinelIntObs.prototype, 'setState')
      const testData1 = ['card 1', 'card 2']
      const testData2 = ['card 3', 'card 4']
      // render
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      const instance = wrapper.getInstance()
      // test with no data
      instance.setState({ fetch: null })
      expect(wrapper).toMatchSnapshot('no data')
      // test on data set 1
      instance.setNewCards(testData1)
      expect(spySetState).toBeCalledWith({ cards: testData1, fetch: null })
      expect(wrapper).toMatchSnapshot('data 1')
      // test on data set 2
      instance.setNewCards(testData2)
      expect(spySetState).toBeCalledWith({ cards: [...testData1, ...testData2], fetch: null })
      expect(wrapper).toMatchSnapshot('data 2')
    })
  })

  describe('#getNewCards', () => {
    let mockSetIsFetching, mockCardFetcher, mockSetNewCards, mockSetState, mockSetEntryCount
    let testCards = ['card1', 'card2', 'card3']

    beforeEach(() => {
      mockSetIsFetching = jest.spyOn(testProps, 'setIsFetching')
      mockSetEntryCount = jest.spyOn(testProps, 'setEntryCount')
      mockCardFetcher = jest.spyOn(testProps, 'cardFetcher').mockImplementation(() => Promise.resolve(testCards))
      mockSetNewCards = jest.spyOn(ScrollerSentinelIntObs.prototype, 'setNewCards').mockReturnValue(null)
      mockSetState = jest.spyOn(ScrollerSentinelIntObs.prototype, 'setState')
    })

    it('is async', async () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      await expect(wrapper.getInstance().getNewCards()).resolves.toBe(undefined)
    })

    it('sets fetching status correctly', async () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      mockSetIsFetching.mockClear()
      await wrapper.getInstance().getNewCards()
      expect(mockSetIsFetching.mock.calls).toEqual([[true], [false]])
    })
    
    it('awaits #cardFetcher', async () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockCardFetcher).toHaveBeenCalled()
    })
    
    it('calls #setNewCards', async () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockSetNewCards).toHaveBeenCalledTimes(1)
      expect(mockSetNewCards).toHaveBeenCalledWith(testCards)
    })

    it('calls #setEntryCount', async () => {
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      const instance = wrapper.getInstance()
      instance.setState({ cards: testCards })
      mockSetEntryCount.mockClear()
      await instance.getNewCards()
      expect(mockSetEntryCount).toHaveBeenCalledTimes(1)
      expect(mockSetEntryCount).toHaveBeenCalledWith(testCards.length)
    })

    it('catches error if #cardFetcher fails', async () => {
      mockCardFetcher.mockRestore()
      mockCardFetcher = jest.spyOn(testProps, 'cardFetcher').mockImplementation(() => Promise.reject('mocked error'))
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockSetState).toHaveBeenCalled()
    })
  })

  describe('#handleIntObs', () => {
    let propsIntObs, mockGetNewCards

    beforeEach(() => {
      propsIntObs = {...testProps}
      propsIntObs.isFetching = false
      mockGetNewCards = jest.spyOn(ScrollerSentinelIntObs.prototype, 'getNewCards').mockReturnValue(null)
    })

    it('should call #getNewCards when all conditions match', () => {
      const wrapper = create(<ScrollerSentinelIntObs {...propsIntObs} />)
      const cb = wrapper.getInstance().handleIntObs()
      const element = document.createElement('article')
      const entries = [new FakeIntersectionObserverEntry(element)]
      cb(entries)
      expect(mockGetNewCards).toHaveBeenCalledTimes(1)
    })

    describe('should not call #getNewCards if', () => {
      test('element is not intersecting', () => {
        jest.spyOn(FakeIntersectionObserverEntry.prototype, 'isIntersecting', 'get')
          .mockReturnValue(false)
        const wrapper = create(<ScrollerSentinelIntObs {...propsIntObs} />)
        const cb = wrapper.getInstance().handleIntObs()
        const element = document.createElement('article')
        const entries = [new FakeIntersectionObserverEntry(element)]
        cb(entries)
        expect(mockGetNewCards).not.toHaveBeenCalled()
      })

      test('instance is fetching', () => {
        propsIntObs.isFetching = true
        const wrapper = create(<ScrollerSentinelIntObs {...propsIntObs} />)
        const cb = wrapper.getInstance().handleIntObs()
        const element = document.createElement('article')
        const entries = [new FakeIntersectionObserverEntry(element)]
        cb(entries)
        expect(mockGetNewCards).not.toHaveBeenCalled()
      })

      test('card.legth >= 200', () => {
        // env setup
        const wrapper = create(<ScrollerSentinelIntObs {...propsIntObs} />)
        const instance = wrapper.getInstance()
        const cb = instance.handleIntObs()
        // Intersection Observer entries
        const element = document.createElement('article')
        const entries = [new FakeIntersectionObserverEntry(element)]
        // cards
        const mockEntry = 'card'
        let cards = new Array(199).fill(mockEntry)

        // 199 entries
        instance.setState({ cards })
        cb(entries)
        expect(mockGetNewCards).toHaveBeenCalledTimes(1)
        // 200 entries
        cards = [mockEntry, ...cards]
        instance.setState({ cards })
        cb(entries)
        expect(mockGetNewCards).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('Intergration', () => {
    it('matches snapshot for happy path', async () => {
      global.IntersectionObserver = FakeIntersectionObserver
      jest.spyOn(ScrollerSentinelIntObs.prototype, 'getNewCards').mockReturnValue([])
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      const cards = [<div>one</div>, <div>two</div>]
      wrapper.getInstance().setState({ fetch: null, cards })
      expect(wrapper).toMatchSnapshot()
    })

    describe('Intersection Observer', () => {
      it('displays loading message on mount if supported', () => {
        global.IntersectionObserver = FakeIntersectionObserver
        jest.spyOn(ScrollerSentinelIntObs.prototype, 'getNewCards').mockReturnValue(['one', 'two'])
        const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
        expect(wrapper).toMatchSnapshot()
      })
  
      it('displays message if not supported', () => {
        global.IntersectionObserver = undefined
        const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
        expect(wrapper).toMatchSnapshot()
      })
    })

    it('displays error message on mount if initial fetch throws error', async () => {
      global.IntersectionObserver = FakeIntersectionObserver
      jest.spyOn(testProps, 'cardFetcher')
        .mockImplementation(() => { throw new Error('mocked error') })
      
      const wrapper = create(<ScrollerSentinelIntObs {...testProps} />)
      expect(wrapper).toMatchSnapshot()
    })
    
    it('displays cards up to point if error thrown by #cardFetcher', async () => {
      global.IntersectionObserver = FakeIntersectionObserver
      const cards = ['one', 'two', 'three']
      jest.spyOn(ScrollerSentinelIntObs.prototype, 'getNewCards')
        .mockImplementation(function() { this.setState({ cards, fetch: null }) })
      const wrapper = mount(<ScrollerSentinelIntObs {...testProps} />)
      const instance = wrapper.instance()

      jest.restoreAllMocks()
      jest.spyOn(testProps, 'cardFetcher').mockImplementation(() => { throw new Error('mocked error') })
      expect(wrapper).toMatchSnapshot()
      
      const html = wrapper.html()
      instance.getNewCards()
      wrapper.update()
      expect(wrapper.html()).toEqual(html)
      wrapper.unmount()
    })
  })
})
