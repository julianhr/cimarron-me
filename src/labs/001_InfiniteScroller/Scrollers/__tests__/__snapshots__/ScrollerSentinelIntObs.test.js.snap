// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollerSentinelIntObs #getCard returns a non-sentinel Card element if datum index != sentinel index 1`] = `
<Card
  description={
    Array [
      "sentence one",
      "sentence two",
    ]
  }
  imgUrl="http://image.com"
  position={14}
  title="test title"
/>
`;

exports[`ScrollerSentinelIntObs #getCard returns a sentinel Card element if datum index == sentinel index 1`] = `
<ForwardRef
  description={
    Array [
      "sentence one",
      "sentence two",
    ]
  }
  forwardedRef={<article />}
  imgUrl="http://image.com"
  position={14}
  title="test title"
/>
`;

exports[`ScrollerSentinelIntObs #setNewCards appends new cards to the current ones: data 1 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
>
  card 1
  card 2
</div>
`;

exports[`ScrollerSentinelIntObs #setNewCards appends new cards to the current ones: data 2 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
>
  card 1
  card 2
  card 3
  card 4
</div>
`;

exports[`ScrollerSentinelIntObs #setNewCards appends new cards to the current ones: no data 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
/>
`;

exports[`ScrollerSentinelIntObs Intergration Intersection Observer displays loading message on mount if supported 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
>
  <div
    className="css-quardp-Message ehwlun20"
  >
    <div>
      Loading...
    </div>
  </div>
</div>
`;

exports[`ScrollerSentinelIntObs Intergration Intersection Observer displays message if not supported 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
>
  <div
    className="css-quardp-Message ehwlun20"
  >
    <div>
      <p>
        This browser doesn't support the Intersection Observer.
      </p>
      <a
        href="https://caniuse.com/#feat=intersectionobserver"
        rel="noopener noreferrer"
        target="_blank"
      >
        Check compatibility
      </a>
    </div>
  </div>
</div>
`;

exports[`ScrollerSentinelIntObs Intergration displays cards up to point if error thrown by #cardFetcher 1`] = `
<Root>
  <div
    className="css-11qniw9-Root e5woe9i0"
  >
    one
    two
    three
  </div>
</Root>
`;

exports[`ScrollerSentinelIntObs Intergration displays error message on mount if initial fetch throws error 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
>
  <div
    className="css-quardp-Message ehwlun20"
  >
    <div>
      Error: mocked error
    </div>
  </div>
</div>
`;

exports[`ScrollerSentinelIntObs Intergration matches snapshot for happy path 1`] = `
<div
  className="css-11qniw9-Root e5woe9i0"
>
  <div>
    one
  </div>
  <div>
    two
  </div>
</div>
`;
